### ISSUE TITLE
-Wunused-variable incorrectly triggering on self-init

### ISSUE THREAD

**Issue by AaronBallman** [llvm/llvm-project#202326]
Consider the following example:

```c
struct [[gnu::warn_unused]] S {
};

void func() {
  struct S s1 = s1;
}
```

Language | Clang | GCC
---| --- | ---
C | No diagnostic | No diagnostic
C++ |  Diagnostic | No diagnostic

https://godbolt.org/z/nafGTonor

I believe Clang's diagnostic in C++ is a false positive; mentioning the variable in its own initialization should silence the diagnostic because the variable is being used.

**Comment by a-tarasyuk** [comment]
Should the warning be omitted only for `S s = s`, `S s(s)`, and `S s = {s}`? Or should more complex self-references also be accepted without a diagnostic?

https://github.com/llvm/llvm-project/blob/2eed192c213322a235197286f08f146b94b20c86/clang/test/SemaCXX/warn-uninitialized-const-reference.cpp#L33

https://github.com/llvm/llvm-project/blob/2eed192c213322a235197286f08f146b94b20c86/clang/test/SemaCXX/uninitialized.cpp#L283

https://github.com/llvm/llvm-project/blob/2eed192c213322a235197286f08f146b94b20c86/clang/test/SemaCXX/uninitialized.cpp#L290

https://github.com/llvm/llvm-project/blob/2eed192c213322a235197286f08f146b94b20c86/clang/test/SemaCXX/uninitialized.cpp#L912-L914

https://github.com/llvm/llvm-project/blob/2eed192c213322a235197286f08f146b94b20c86/clang/test/SemaCXX/uninitialized.cpp#L916

**Comment by AaronBallman** [comment]
> Should the warning be omitted only for `S s = s`, `S s(s)`, and `S s = {s}`?

I think any mention of the variable should silence the diagnostic: https://godbolt.org/z/8jf5aW9re

> Or should more complex self-references also be accepted without a diagnostic?
> 
> [llvm-project/clang/test/SemaCXX/warn-uninitialized-const-reference.cpp](https://github.com/llvm/llvm-project/blob/2eed192c213322a235197286f08f146b94b20c86/clang/test/SemaCXX/warn-uninitialized-const-reference.cpp#L33)
> 
> Line 33 in [2eed192](/llvm/llvm-project/commit/2eed192c213322a235197286f08f146b94b20c86)
>  A a2 = const_use_A(a2);       // expected-warning {{variable 'a2' is uninitialized when used within its own initialization}} 
> 
> [llvm-project/clang/test/SemaCXX/uninitialized.cpp](https://github.com/llvm/llvm-project/blob/2eed192c213322a235197286f08f146b94b20c86/clang/test/SemaCXX/uninitialized.cpp#L283)
> 
> Line 283 in [2eed192](/llvm/llvm-project/commit/2eed192c213322a235197286f08f146b94b20c86)
>  A a14 = A(a14);  // expected-warning {{variable 'a14' is uninitialized when used within its own initialization}} 
> 
> [llvm-project/clang/test/SemaCXX/uninitialized.cpp](https://github.com/llvm/llvm-project/blob/2eed192c213322a235197286f08f146b94b20c86/clang/test/SemaCXX/uninitialized.cpp#L290)
> 
> Line 290 in [2eed192](/llvm/llvm-project/commit/2eed192c213322a235197286f08f146b94b20c86)
>  A a21 = {a21};  // expected-warning {{variable 'a21' is uninitialized when used within its own initialization}} 
> 
> [llvm-project/clang/test/SemaCXX/uninitialized.cpp](https://github.com/llvm/llvm-project/blob/2eed192c213322a235197286f08f146b94b20c86/clang/test/SemaCXX/uninitialized.cpp#L912-L914)
> 
> Lines 912 to 914 in [2eed192](/llvm/llvm-project/commit/2eed192c213322a235197286f08f146b94b20c86)
>  A a1([=] { // expected-warning{{variable 'a1' is uninitialized when used within its own initialization}} 
>    return a1.x; 
>  }); 
> 
> [llvm-project/clang/test/SemaCXX/uninitialized.cpp](https://github.com/llvm/llvm-project/blob/2eed192c213322a235197286f08f146b94b20c86/clang/test/SemaCXX/uninitialized.cpp#L916)
> 
> Line 916 in [2eed192](/llvm/llvm-project/commit/2eed192c213322a235197286f08f146b94b20c86)
>  A a3([=] { return a3.x; }()); // expected-warning{{variable 'a3' is uninitialized when used within its own initialization}}

I don't think this should impact diagnostics other than `-Wunused-variable`; those are all controlled by `-Wuninitialized` right? I don't think that behavior would change.
